### PR TITLE
Vectorized execution support for disk-spilling operators

### LINKED ISSUE #5968: [SH] Vectorization: Add Disk Spilling to Operators
Background: https://airtable.com/tblD3oZPLJgGhCmch/viw1DKmbKhg2MIECH/recagjuRqKN5Jxh1n

Description: This work allows the SQL Vectorization Engine to work will the majority of SQL statements including buffering operators such as hash join, hash aggregation, and global sort.

Increased query execution performance for all queries including OLTP heavy queries such as aggregations. Disk spilling is needed to process the remaining queries not yet covered by vectorized execution. Disk spilling occurs when certain queries consume too much memory; CockroachDB will spill parts of the memory needed for query execution to disk to avoid consuming all of the available working memory and allow other queries to continue to make progress.

Team: Andrew Woods, Jordan Lewis, Alfonso Subiotto Marques, Yahor Yuzefovich

Github Tracking Issue: https://github.com/cockroachdb/cockroach/issues/36507 https://github.com/cockroachdb/docs/issues/5968

### LINKED ISSUE #6203: colexec: add memory accounting logic to Allocator
PR: https://github.com/cockroachdb/cockroach/pull/42468

From release notes:
> When executed via the [vectorized engine](../v20.1/vectorized-execution.html), make each buffering operator use at most `sql.distsql.temp_storage.workmem` memory (which is 64MB by default).  Previously, all buffering operators (like hash and merge joins, sorts) could use arbitrary amounts of memory which could lead to OOM panics. [#42468][#42468] {% comment %}doc{% endcomment %}

### LINKED ISSUE #6618: colexec: add unordered distinct
PR: https://github.com/cockroachdb/cockroach/pull/42522

From release notes:
> [Vectorized execution engine](../v20.1/vectorized-execution.html) now supports `DISTINCT` on unordered input. [#42522][#42522] {% comment %}doc{% endcomment %}

### LINKED ISSUE #6621: colexec: add bool_or/bool_and operators
PR: https://github.com/cockroachdb/cockroach/pull/44164

From release notes:
> The [vectorized execution engine](../v20.1/vectorized-execution.html) now supports `bool_and`/`bool_or` [builtin](../v20.1/functions-and-operators.html) aggregation functions. [#44164][#44164] {% comment %}doc{% endcomment %}

### LINKED ISSUE #6905: colexec: plan disk-spilling enabled operators when vectorize=auto
PR: https://github.com/cockroachdb/cockroach/pull/45582

From release notes:
> Hash joins and sorts are now run using the vectorized engine when `vectorize=auto` (default configuration). [#45582][#45582] {% comment %}doc{% endcomment %}

### LINKED ISSUE #6922: colexec: add disk spilling to merge joiner
PR: https://github.com/cockroachdb/cockroach/pull/45784

From release notes:
> Queries with [`MERGE` join](../v20.1/joins.html) can now run via the vectorized engine when `vectorize` is set to `auto`. [#45784][#45784] {% comment %}doc{% endcomment %}

### LINKED ISSUE #6924: colexec: enable hash aggregator to run in vecotrize auto
PR: https://github.com/cockroachdb/cockroach/pull/45832

From release notes:
> Hash aggregator is now supported in `vectorize=auto` mode. [#45832][#45832] {% comment %}doc{% endcomment %}

### LINKED ISSUE #6979: sql: make some tweaks to using the vectorized engine
PR: https://github.com/cockroachdb/cockroach/pull/46080

From release notes:
> `experimental_on` option for `vectorize` [session variable](../v20.1/set-vars.html) has been renamed to `on`. The only things that will not run with `auto`, but will run with `on`, are unordered [`DISTINCT`](../v20.1/select-clause.html#eliminate-duplicate-rows) and two [window functions](../v20.1/window-functions.html) (`percent_rank` and `cume_dist`). The two options are otherwise identical. [#46080][#46080] {% comment %}doc{% endcomment %}

### LINKED ISSUE #7141: release-20.1: colexec: rename vectorize setting 192auto to auto and make it the default
PR: https://github.com/cockroachdb/cockroach/pull/46925

From release notes:
> The [vectorized execution engine](../v20.1/vectorized-execution.html) now only runs queries with streaming operators. To enable vectorized execution for buffering operators, use `SET vectorize=on`. [#46925][#46925] {% comment %}doc{% endcomment %}

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes #5968.
Fixes #6203. 
Fixes #6922. 
Fixes #6924.
Fixes #6905.
Fixes #6618.
Fixes #6621.
Fixes #6979.
Fixes #7141.

- Updated `vectorize` settings on 20.1 Vectorized Execution doc page and Session Variables table.
- Added notes to Window Functions and Selection Queries pages about experimental support in vectorized.
- Updated 20.1 Vectorized Execution page for clarity, accuracy around disk-spilling support.
- Updated 19.2 Vectorized Execution page for clarity.

### CODE REVIEW

**Comment by ericharmeling** [comment]
@yuzefovich 
Please review latest updates. Thanks!

File: v20.1/vectorized-execution.md [hunk 1]
```diff
@@ -8,20 +8,28 @@ CockroachDB supports [column-oriented](https://en.wikipedia.org/wiki/Column-orie
 
 Many SQL databases execute [query plans](https://en.wikipedia.org/wiki/Query_plan) one row of table data at a time. Row-oriented execution models can offer good performance for [online transaction processing (OLTP)](https://en.wikipedia.org/wiki/Online_transaction_processing) queries, but suboptimal performance for [online analytical processing (OLAP)](https://en.wikipedia.org/wiki/Online_analytical_processing) queries. The CockroachDB vectorized execution engine dramatically improves performance over [row-oriented execution](https://en.wikipedia.org/wiki/Column-oriented_DBMS#Row-oriented_systems) by processing each component of a query plan on type-specific batches of column data.
 
+{{site.data.alerts.callout_success}}
+CockroachDB does not support vectorized execution for all data types. For details, see [supported data types](#supported-data-types).
+{{site.data.alerts.end}}
+
 ## Configuring vectorized execution
```
Comment (awoods187) [inline v20.1/vectorized-execution.md:16]: we could say "By default, `vectorize=auto`, xxx"

File: v20.1/vectorized-execution.md [hunk 2]
```diff
@@ -8,20 +8,28 @@ CockroachDB supports [column-oriented](https://en.wikipedia.org/wiki/Column-orie
...
+You can turn vectorized execution on or off for all queries in the current session with the `vectorize` [session variable](set-vars.html). The following options are supported:
 
-Option | Description
+Option    | Description
 ----------|------------
-`auto` | Instructs CockroachDB to use the vectorized execution engine on operations that always execute in memory, without the need to spill intermediate results to disk.<br><br>**Default:** `vectorize=auto`
-`experimental_on` | Turns on vectorized execution for all operations. We do not recommend using this option in production environments, as it can lead to memory issues.<br/>See [Known Limitations](#known-limitations) for more information.
-`off` | Turns off vectorized execution for all operations.
+`auto` | Instructs CockroachDB to use the vectorized execution engine on most queries that execute in memory, without the need to [spill intermediate results to disk](#disk-spilling-operations).<br><br>**Default:** `vectorize=auto`
+`on`   | Turns on vectorized execution for all queries.
```
Comment (awoods187) [inline v20.1/vectorized-execution.md:24]: maybe say "for all supported queries"

File: v20.1/vectorized-execution.md [hunk 3]
```diff
@@ -46,11 +54,23 @@ For information about vectorized execution in the context of the CockroachDB arc
 
 For detailed examples of vectorized query execution for hash and merge joins, see the blog posts [40x faster hash joiner with vectorized execution](https://www.cockroachlabs.com/blog/vectorized-hash-joiner/) and [Vectorizing the merge joiner in CockroachDB](https://www.cockroachlabs.com/blog/vectorizing-the-merge-joiner-in-cockroachdb/).
 
-## Known limitations
+## Disk-spilling operations
+
+By default, vectorized execution is disabled for the following memory-intensive operations:
+
+- Global [sorts](query-order.html).
```
Comment (awoods187) [inline v20.1/vectorized-execution.md:61]: nit: i don't think we typically use periods at the ends of bullets

File: v20.1/vectorized-execution.md [hunk 4]
```diff
@@ -69,6 +89,10 @@ Vectorized execution is supported for the following [data types](data-types.html
 CockroachDB uses the vectorized engine to execute queries on columns with supported data types, even if a column's parent table includes unused columns with unsupported data types.
 {{site.data.alerts.end}}
 
+## Known limitations
+
+Vectorized execution is not as extensively tested as CockroachDB's existing row-oriented execution engine. In addition, [some data types are not supported](#supported-data-types), and some queries are not fully supported.
```
Comment (awoods187) [inline v20.1/vectorized-execution.md:94]: I don't think we should say >Vectorized execution is not as extensively tested as CockroachDB's existing row-oriented execution engine

It is tested just as much as the rest of CRDB. Let's delete this line.

**Comment by awoods187** [comment]
@dbist PTAL at the updated language here from the product council

**Pull request merged**